src/common/functions/src/aggregates/aggregate_distinct_state.rs

```diff
@@ -0,0 +1,395 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::RandomState;
+use std::collections::HashSet;
+use std::convert::TryFrom;
+
+use bytes::BytesMut;
+use common_arrow::arrow::bitmap::Bitmap;
+use common_datavalues::prelude::*;
+use common_datavalues::DFTryFrom;
+use common_exception::Result;
+use common_io::prelude::*;
+use ordered_float::OrderedFloat;
+use serde::Deserialize;
+use serde::Serialize;
+
+use super::aggregate_distinct_state::DataGroupValue;
+
+pub trait DistinctStateFunc<S>: Send + Sync {
+    fn new() -> Self;
+    fn serialize(&self, writer: &mut BytesMut) -> Result<()>;
+    fn deserialize(&mut self, reader: &mut &[u8]) -> Result<()>;
+    fn is_empty(&self) -> bool;
+    fn len(&self) -> usize;
+    fn add(&mut self, columns: &[ColumnRef], row: usize) -> Result<()>;
+    fn batch_add(
+        &mut self,
+        columns: &[ColumnRef],
+        validity: Option<&Bitmap>,
+        input_rows: usize,
+    ) -> Result<()>;
+    fn merge(&mut self, rhs: &Self) -> Result<()>;
+    fn build_columns(&self, fields: &[DataField]) -> Result<Vec<ColumnRef>>;
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq, Hash, Clone)]
+pub struct DataGroupValues(Vec<DataGroupValue>);
+
+pub struct AggregateDistinctState {
+    set: HashSet<DataGroupValues, RandomState>,
+}
+pub struct AggregateDistinctIntegerState<T: PrimitiveType + std::hash::Hash + Eq> {
```

Comment:
Can we combine `AggregateDistinctIntegerState ` and `AggregateDistinctFloatState `  into one struct:

```
pub struct AggregatePrimitiveState<T: PrimitiveType, E: From<T>> {
    set: HashSet<E, RandomState>,
    _t: PhantomData<T>,
}
```